pkg/services/shipment_summary_worksheet/shipment_summary_worksheet.go

```
@@ -397,11 +402,36 @@ func formatSignedCertifications(signedCertifications []*models.SignedCertificati
 }
 
-// FormatSignatureDate formats the date the office members signed the SSW
-func FormatSignatureDate(signature time.Time) string {
-	dateLayout := "02 Jan 2006" // Removed time to save space on template, per PO it's not needed
-	dt := signature.Format(dateLayout)
-	return dt
+// The following formats the preparation date, as the preparation date for AOAs is the date the service counselor certifies the advance.
+func formatAOADate(signedCertifications []*models.SignedCertification, ppmid uuid.UUID) string {
+	// This loop evaluates certs to find Office AOA Signature date
+	for _, cert := range signedCertifications {
+		if cert.PpmID != nil { // Required to avoid error, service members signatures have nil ppm ids
+			if *cert.PpmID == ppmid { // PPM ID needs to be checked to prevent signatures from other PPMs on the same move from populating
+				if *cert.CertificationType == models.SignedCertificationTypePreCloseoutReviewedPPMPAYMENT {
+					aoaDate := FormatDate(cert.UpdatedAt) // We use updatedat to get the most recent signature dates
+					return aoaDate
+				}
+			}
+		}
+	}
+	return FormatDate(time.Now())
+}
+
+// The following formats the preparation date, as the preparation date for SSWs is the date the closeout counselor certifies the closeout.
+func formatSSWDate(signedCertifications []*models.SignedCertification, ppmid uuid.UUID) string {
+	// This loop evaluates certs to find Office SSW Signature date
+	for _, cert := range signedCertifications {
+		if cert.PpmID != nil { // Required to avoid error, service members signatures have nil ppm ids
+			if *cert.PpmID == ppmid { // PPM ID needs to be checked to prevent signatures from other PPMs on the same move from populating
+				if *cert.CertificationType == models.SignedCertificationTypeCloseoutReviewedPPMPAYMENT {
+					sswDate := FormatDate(cert.UpdatedAt) // We use updatedat to get the most recent signature dates
+					return sswDate
+				}
+			}
+		}
+	}
+	return "Uncertified"
```

Lines 418 and 434: I'm assuming this is a worse case scenario?  It should never get to this,  right? You  may want to throw an error and/or log here.

---

418 is what we discussed in the standup this morning. Right now, there's a way to download the AOA before the SC submits move details where their signature is added. Until we can revisit the aoa download availability, our PO just wants it to use the date the PDF generation was ran.

As for 434, this should never occur, so an error may be more appropriate here. If that's what we want to go with, I'll add it to the todo list for later today.

---

I've gone ahead and added the error handling to this method. Now, if there's ever an uncertified payment packet downloaded before it should be, it will error out. Honestly, making this sort of thing easier in this service was long overdue, so I think that's a great change.

As for the generic method idea, I'm going to push back against that because that's the solution I attempted initially. Due to the implementation in the way that the service determines which packet is being generated, as well as the best way for handling adding the dates themselves from the certs, and the two alternate solutions to when a date shouldn't be used... I believe the current solution is easier to read, easier to maintain, is marginally better performance-wise, and uses about the same amount of code as combining them. 

I do realize that the amount of methods in this service is... kind of a lot, so I'd see why we'd want to cut down on them. But if I need more justification still, then an additional reason for adding 2 in this story is that these changes also removed a near-duplicate method in formatSignatureDate.

Just let me know if any objections.